File: strategy/marvin/group/move/safecorner.lua
local SafeCorner = Class("Group.Move.SafeCorner", require "group/move/base")

local Referee = require "../base/referee"
local World = require "../base/world"
local Freekick = require "agent/attacker/freekick"
local MoveToPos = require "task/movetopos"
local StopAttack = require "task/stopattack"
local Striker = require "task/striker"
local G = World.Geometry

SafeCorner.N_ROBOTS = 5

function SafeCorner.canStart()
	return  World.Ball.pos.y > 4 * G.FieldHeightHalf / 5 --and Referee.opponentTouchedLast()
		and math.abs(World.Ball.pos.x) > G.FieldWidthHalf / 2
		and World.RefereeState == "Stop"
end

function SafeCorner:_init()
	self._ballSide = (World.Ball.pos.x > 0) and 1 or -1 --Instanzvariable
	self._goalDist = G.DefenseRadius + 0.4
end

function SafeCorner:_canContinue()
	if Referee.isFriendlyFreeKickState() then
		return true
	end
	return World.Ball.pos.y > 4 * G.FieldHeightHalf / 5 - 0.2	--Eckposition festlegen
		and math.abs(World.Ball.pos.x) > G.FieldWidthHalf / 2 - 0.2 --G: geometry
		and World.RefereeState == "Stop"
end

function SafeCorner:_updateTasks()

	local taskAssignments = {}
	if World.RefereeState == "Stop" then
		taskAssignments[self._robots[1]] = { class = StopAttack, params = { } }
	elseif Referee.isFriendlyFreeKickState() then
		taskAssignments[self._robots[1]] = { behavior = Freekick }
	end

	taskAssignments[self._robots[2]] = { class = Striker, params = { Vector(0, G.FieldHeightHalf * -0.5), Vector(0, 0) }}
	taskAssignments[self._robots[3]] = { class = Striker, params = { Vector(self._ballSide * G.FieldWidthHalf * -0.5, G.FieldHeightHalf * -0.5),
		Vector(self._ballSide * G.FieldWidthHalf * -0.5, 0) }}
	taskAssignments[self._robots[4]] = { class = MoveToPos, params = { Vector(0.3, G.OpponentGoal.y - G.DefenseRadius - 0.4)}}
	-- taskAssignments[self._robots[5]] = { class = MoveToPos, params = { Vector(, )}}


	return taskAssignments, self._robots[1]
end
return SafeCorner

File: strategy/typescript/glados/group/move/safecorner.ts
let SafeCorner = Class("Group.Move.SafeCorner", require "group/move/base")

import * as Referee from "base/referee";
import * as World from "base/world";
let Freekick = require "agent/attacker/freekick"
import {MoveToPos} from "glados/task/shared/movetopos";
let StopAttack = require "task/attacker/stopattack"
import {Striker} from "glados/task/attacker/striker";
let G = World.Geometry

SafeCorner.MIN_ROBOTS = 5
SafeCorner.MAX_ROBOTS = 5

function SafeCorner.canStart () {
	return  World.Ball.pos.y > 4 * G.FieldHeightHalf / 5 //and Referee.opponentTouchedLast()
		 &&  Math.abs(World.Ball.pos.x) > G.FieldWidthHalf / 2
		 &&  World.RefereeState == "Stop"
}

function SafeCorner:_init () {
	this._ballSide = (World.Ball.pos.x > 0) ? 1 : -1 //Instanzvariable
	this._goalDist = G.DefenseRadius + 0.4
}

function SafeCorner:_canContinue () {
	if (Referee.isFriendlyFreeKickState()) {
		return true
	}
	return World.Ball.pos.y > 4 * G.FieldHeightHalf / 5 - 0.2 //Eckposition festlegen
		 &&  Math.abs(World.Ball.pos.x) > G.FieldWidthHalf / 2 - 0.2 //G: geometry
		 &&  World.RefereeState == "Stop"
}

function SafeCorner:_updateTasks () {

	let taskAssignments = {}
	if (World.RefereeState == "Stop") {
		taskAssignments[this._robots[0]] = { class: StopAttack, params: { } }
	} else if (Referee.isFriendlyFreeKickState()) {
		taskAssignments[this._robots[0]] = { behavior: Freekick }
	}

	taskAssignments[this._robots[1]] = { class: Striker, params: { Vector(0, G.FieldHeightHalf * -0.5), new Vector(0, 0) }}
	taskAssignments[this._robots[2]] = { class: Striker, params: { Vector(this._ballSide * G.FieldWidthHalf * -0.5, G.FieldHeightHalf * -0.5),
		Vector(this._ballSide * G.FieldWidthHalf * -0.5, 0) }}
	taskAssignments[this._robots[3]] = { class: MoveToPos, params: { Vector(0.3, G.OpponentGoal.y - G.DefenseRadius - 0.4)}}
	// taskAssignments[this._robots[4]] = { class: MoveToPos, params: { Vector(, )}}


	return taskAssignments, this._robots[0]
}
return SafeCorner
